src/Butler/Auth/OIDC.hs

```diff
@@ -0,0 +1,206 @@
+
+data OIDCProviderConfig = OIDCProviderConfig
+    { opIssuerURL :: Text
+    , opClientID :: Text
+    , opClientSecret :: Text
+    , opAppPublicURL :: Text
+    , opUserClaim :: Maybe Text
+    , opEnforceAuth :: Bool
+    , opName :: Text
+    }
+data OIDCEnv = OIDCEnv
+    { oidc :: O.OIDC
+    , manager :: Manager
+    , provider :: O.Provider
+    , redirectUri :: ByteString
+    , sessionStoreStorage :: MVar (HM.Map O.State O.Nonce)
+    , providerConfig :: OIDCProviderConfig
+    }
+
+initOIDCEnv :: OIDCProviderConfig -> IO OIDCEnv
+initOIDCEnv providerConfig@OIDCProviderConfig{..} = do
+    manager <- newTlsManager
+    provider <- O.discover opIssuerURL manager
+    sessionStoreStorage <- newMVar HM.empty
+    let redirectUri = encodeUtf8 opAppPublicURL <> "_cb"
+        clientId = encodeUtf8 opClientID
+        clientSecret = encodeUtf8 opClientSecret
+        oidc = O.setCredentials clientId clientSecret redirectUri (O.newOIDC provider)
+    pure OIDCEnv{..}
+
+data OIDCState = OIDCState
+    { randomT :: Text
+    , uri :: Maybe Text
+    }
+    deriving (Generic, Show)
+
+instance FromJSON OIDCState
+
+instance ToJSON OIDCState
+
+-- | Generate a random fixed size string of 42 char base64 encoded
+genRandomB64 :: IO ByteString
+genRandomB64 = B64.encode . hash <$> genRandom
+
+-- | Generate a random fixed size string of 1024 Bytes
+genRandom :: IO ByteString
+genRandom = do
+    g <- Random.newStdGen
+    let (bs, _ng) = genByteString 1024 g
+    pure bs
+
+mkSessionStore :: OIDCEnv -> Maybe O.State -> Maybe Text -> O.SessionStore IO
+mkSessionStore OIDCEnv{sessionStoreStorage} stateM uriM = do
+    let sessionStoreGenerate = do
+            rb <- liftIO genRandomB64
+            let s = OIDCState (decodeUtf8 rb) uriM
+            pure (B64.encode $ BSL.toStrict $ Data.Aeson.encode s)
+        sessionStoreSave = storeSave
+        sessionStoreGet = storeGet
+        sessionStoreDelete = case stateM of
+            Just state' -> modifyMVar_ sessionStoreStorage $ \store -> pure $ HM.delete state' store
+            Nothing -> pure ()
+     in O.SessionStore{..}
+  where
+    storeSave :: O.State -> O.Nonce -> IO ()
+    storeSave state' nonce = modifyMVar_ sessionStoreStorage $ \store -> pure $ HM.insert state' nonce store
+    storeGet :: O.State -> IO (Maybe O.Nonce)
+    storeGet state' = do
+        store <- readMVar sessionStoreStorage
+        let nonce = HM.lookup state' store
+        pure nonce
+
+loginServer :: OS -> Process -> Sessions -> (Html () -> Html ()) -> JWTSettings -> AuthResult SessionID -> OIDCEnv -> Maybe Workspace -> ServerT LoginAPI Servant.Handler
+loginServer os process sessions mkIndexHtml jwtSettings auth oidcenv mWorkspace =
+    indexRoute auth :<|> loginRoute oidcenv :<|> callbackRoute oidcenv :<|> guestCallbackRoute
+  where
+    callbackRoute :: OIDCEnv -> Maybe Text -> Maybe Text -> Maybe Text -> Servant.Handler AuthResp
+    callbackRoute oidcEnv errM codeM stateM = do
```

Comment:
The 'M' suffix is often used for action variable, it's more common to use the 'm' prefix for maybe variable, e.g.: 
```suggestion
    callbackRoute oidcEnv mErr mCode mState = do
```